Documents/config.md
###Config

I. Create object for config:
```js
var env = {
        /**
         * Specify environment variables:
         * ENV_TYPE - type of environment to use. It must be one of custom configs
         */
//      ENV_TYPE: "custom",
        URL_SSO: "http://sso.devportal-qa.dspdev.wmg.com",
        URL_DEVPORTAL: "http://security-console.devportal-ci.dspdev.wmg.com/",
        ENV_TYPE: "custom"
    },
    config = {
        /**
         * This part will be always in config
         */
        _default_: {
            client_id: "devportal",
            client_secret: "appclientsecret",
            user_name: "DSP_TestUser2",
            user_password: "W@rn3rTestU$3r"    
        },
        
        /**
         * Add environment variables to the config
         */
        _env_: env,
        
        /**
         * Custom configs: "qa" and "custom". 
         * If you want to use "qa" settings, specify `_env_ -> ENV_TYPE` as "qa"
         * If you want to use "custom" settings, specify `_env_ -> ENV_TYPE` as "custom"
         *      And don't forget to specify 
         */
        qa: {
            urls: {
                sso: "http://sso.devportal-qa.dspdev.wmg.com",
                security_console: "http://security-console.devportal-ci.dspdev.wmg.com/"
            },
            user_name: "uaa_test_user01@wmgdsp.dev",
            user_password: "No!daIN@124"
        },
        custom: {
            urls: {
                sso: env.URL_SSO,
                security_console: env.URL_DEVPORTAL
            }
        }
    };
```

II. Creating config:

```js
config = ftf.configer.process(config);
```
You should pass your config object to configer factory. As a result you should get object with next structure (if using the example from this document):

```js
{ 
    client_id: 'devportal',
    client_secret: 'appclientsecret',
    user_name: 'uaa_test_user01@wmgdsp.dev',
    user_password: 'No!daIN@124',
    urls: { 
        sso: 'http://sso.devportal-qa.dspdev.wmg.com',
        security_console: 'http://security-console.devportal-ci.dspdev.wmg.com/' 
    },
    _cli: {
        //cli parameters should be here
    },
    _env: {
        //environment variables should be here
    }
}
```


Documents/config.js.md
##conf/config.js

###A. Create object for config:

####I. Initializing configer module of Framework

```js
var configer = ftf.configer,
```

####II. Processing Command Line arguments

```js
    cli = configer.getParamsFromCli(),
```

####III. Creating Object with Environment Variables

```js
    env = {
        URL_SSO: cli["URL_SSO"] || "http://sso.devportal-qa.dspdev.wmg.com",
        URL_DEVPORTAL: cli["URL_DEVPORTAL"] || "http://security-console.devportal-ci.dspdev.wmg.com/",
        ENV_TYPE: cli["ENV_TYPE"] || "ci"
    },
```

####IV. Starting Creating Config pre-object

```js
    config = {
```

####V. Default Values:

```js
        _default_: {
            client_id: "devportal",
            client_secret: "appclientsecret",
            user_name: "DSP_TestUser2",
            user_password: "W@rn3rTestU$3r",
        },
```

####VI. System values:

```js
        _system_: {
            browser: (cli.browser in ["chrome", "firefox", "ie"] ? cli.browser : "chrome"),
            resolution: {
                width: 800,
                height: 600
            },
            reporting: cli.reporting in ["html", "xml"] ? cli.reporting : "none"
        },
```

####VIII. Adding Object with Environment Variables to pre-config object

```js
        _env_: env,
```

####IX. Setting up different environment set ups for executing. If variables from environment set up are in conflicts with variables from `_default_`, variables from `_default_` are overriden by variables from environment set up.

```js
        ci: {
            urls: {
                sso: "http://sso.devportal-ci.dspdev.wmg.com",
                security_console: "http://security-console.devportal-ci.dspdev.wmg.com/"
            }
        },
        qa: {
            urls: {
                sso: "http://sso.devportal-qa.dspdev.wmg.com",
                security_console: "http://security-console.devportal-ci.dspdev.wmg.com/"
            },
            user_name: "uaa_test_user01@wmgdsp.dev",
            user_password: "No!daIN@124"
        },
        localhost: {
             urls: {
                 sso: "http://sso.devportal-ci.dspdev.wmg.com",
                 security_console:  "http://localhost:9000"
             }
        },
        custom: {
            urls: {
                sso: env.URL_SSO,
                security_console: env.URL_DEVPORTAL
            }
        }
    };
```

####X. Compiling config object and exporting it.

```js
config = configer.process(config);
module.exports = config;
```

###B. Creating config: 
(See conf/protractor-conf.js.md file)

```js
config = ftf.configer.process(config);
```
You should pass your config object to configer factory. As a result you should get object with next structure (if using the example from this document):

```js
{ 
    client_id: 'devportal',
    client_secret: 'appclientsecret',
    user_name: 'uaa_test_user01@wmgdsp.dev',
    user_password: 'No!daIN@124',
    _system_: {
        browser: 'chrome',
        resolution:  {
            width: 800,
            height: 600
        },
        reporting: 'xml'
    },
    urls: { 
        sso: 'http://sso.devportal-qa.dspdev.wmg.com',
        security_console: 'http://security-console.devportal-ci.dspdev.wmg.com/' 
    },
    _cli: {
        //cli parameters should be here
    },
    _env: {
        //environment variables should be here
    }
}
```


login_test/conf/config.js
var configer = ftf.configer,
    cli = configer.getParamsFromCli(),
    env = {
        URL_SSO: cli["URL_SSO"] || "http://sso.devportal-ci.dspdev.wmg.com",
        URL_DEVPORTAL: cli["URL_DEVPORTAL"] || "http://security-console.devportal-ci.dspdev.wmg.com/",
        ENV_TYPE: cli["ENV_TYPE"] || "ci"
    },
    config = {
        _default_: {
            client_id: "devportal",
            client_secret: "appclientsecret",
            user_name: configer.getEnvVarByKey("USER_NAME") || "DSP_TestUser2",
            user_password: configer.getEnvVarByKey("USER_PASSWORD") || "W@rn3rTestU$3r"
        },
        _system_: {
            browser: (cli.browser in ["chrome", "firefox", "ie"] ? cli.browser : "chrome"),
            resolution: {
                width: 1200,
                height: 1000
            },
            reporting: cli.reporting in ["html", "xml"] ? cli.reporting : "none"
        },
        _env_: env,
        ci: {
            urls: {
                sso: "http://sso.devportal-ci.dspdev.wmg.com",
                app_url: "http://security-console.devportal-ci.dspdev.wmg.com/"
            }
        },
        qa: {
            urls: {
                sso: "http://sso.devportal-qa.dspdev.wmg.com",
                app_url: "http://security-console.devportal-qa.dspdev.wmg.com/"
            },
            user_name: "uaa_test_user01@wmgdsp.dev",
            user_password: "No!daIN@124"
        },
        localhost: {
             urls: {
                 sso: "http://sso.devportal-ci.dspdev.wmg.com",
                 app_url:  "http://localhost:9000"
             }
        },
        custom: {
            urls: {
                sso: env.URL_SSO,
                app_url: env.URL_DEVPORTAL
            }
        }
    };
    config = configer.process(config);
    
module.exports = config;